File: backend/src/controllers/api/bars.ts
import { Request, Response } from "express";
import { getBars } from "../../providers/index.js";
import { Ticker, Frequency } from "@shared/types";

export async function barsController(req: Request, res: Response) {
  try {
    const { symbols, freq, start, end } = req.query;

    if (!symbols || !freq || !start || !end) {
      res.status(400).json({ error: "Missing required query params: symbols, freq, start, end." });
      return;
    }

    const tickerList: Ticker[] = (symbols as string).split(",").map((s) => ({ symbol: s.trim() }));
    const frequency = freq as Frequency;
    const startDate = new Date(start as string);
    const endDate = new Date(end as string);

    // Always respond with the { [symbol]: OHLCVData[] } shape, regardless of
    // how many symbols were requested, so clients don't need to special-case
    // the single-symbol response.
    const bars = await getBars(tickerList, frequency, startDate, endDate);
    res.json(bars);
  } catch (error) {
    console.error("Error fetching bars:", error);
    res.status(500).json({ error: "Failed to fetch bars." });
  }
}

File: backend/app/routers/bars.py
import logging
from datetime import datetime

from fastapi import APIRouter, HTTPException

from app.providers import alpaca
from app.schemas.stock import Frequency, OHLCVData, Ticker

logger = logging.getLogger(__name__)
router = APIRouter()


@router.get("/bars")
async def get_bars(
    symbols: str,
    freq: Frequency,
    start: datetime,
    end: datetime,
) -> dict[str, list[OHLCVData]]:

    ticker_list = [Ticker(symbol=s.strip()) for s in symbols.split(",") if s.strip()]
    if not ticker_list:
        raise HTTPException(status_code=400, detail="No valid symbols provided.")

    try:
        return await alpaca.get_bars(ticker_list, freq, start, end)
    except Exception as e:
        logger.exception("Failed to fetch bars from Alpaca")
        raise HTTPException(
            status_code=502, detail="Failed to fetch bars from Alpaca."
        ) from e
